### ISSUE TITLE
[Build] Eigen hash mismatch

### ISSUE THREAD

**Issue by smartnet-club** [microsoft/onnxruntime#18286]
### Describe the issue

Error building rel-1.16.1

### Target platform

Ubuntu 22.04

### Build script

`build.sh --config ${ONNXRUNTIME_CONFIG} --allow_running_as_root --parallel --build_shared_lib --build_wheel --use_extensions --use_dnnl --use_openvino ${DEVICE} --cuda_home /usr/local/cuda --cudnn_home /usr/lib/x86_64-linux-gnu/ --use_tensorrt --tensorrt_home /usr/lib/x86_64-linux-gnu/ --skip_tests --skip_submodule_sync --cmake_extra_defines CMAKE_CUDA_ARCHITECTURES=${CMAKE_CUDA_ARCHITECTURES}
`

### Error / output

19.51 -- Using src='https://gitlab.com/libeigen/eigen/-/archive/3.4/eigen-3.4.zip'
20.11 -- verifying file...
20.11        file='/usr/src/onnxruntime/build/Linux/Release/_deps/eigen-subbuild/eigen-populate-prefix/src/eigen-3.4.zip'
20.12 -- SHA1 hash of
20.12     /usr/src/onnxruntime/build/Linux/Release/_deps/eigen-subbuild/eigen-populate-prefix/src/eigen-3.4.zip
20.12   does not match expected value
20.12     expected: 'ee201b07085203ea7bd8eb97cbcb31b07cfa3efb'
20.12       actual: '123c29af62216a70611d8b826c0e57c361c9a49f'
20.12 -- Hash mismatch, removing...

**Comment by smartnet-club** [comment]
The problem is you use meta package 'https://gitlab.com/libeigen/eigen/-/archive/3.4/eigen-3.4.zip'

**Comment by chitoku** [comment]
This line was added in `cmake/deps.txt` on Sept 8th, 
https://github.com/microsoft/onnxruntime/commit/bc84f52633444a2fef336f6eeb3113a8a5c19768#diff-12c22e06cbb37ea0ed9f9eaf60cbe408dbeef04072df6a9f431c3290822ea835
```
eigen;https://gitlab.com/libeigen/eigen/-/archive/3.4/eigen-3.4.zip;ee201b07085203ea7bd8eb97cbcb31b07cfa3efb
```

**Comment by edwintorok** [comment]
> The problem is that you are using the metapackage 'https://gitlab.com/libeigen/eigen/-/archive/3.4/eigen-3.4.zip'.
> 
> To have a stable hash you need to use the exact version of 'https://gitlab.com/libeigen/eigen/-/archive/3.4.0/eigen-3.4.0.zip'.

I've opened a PR now that does that and it fixes the build for me. (`3.4` refers to a branch, not a tag, the tag is `3.4.0` as you said, and there is now 3.4.1 content on the 3.4 branch which is why the hash has change and the build here was broken).

**Comment by smartnet-club** [comment]
@edwintorok could you fix [latest release branch](https://github.com/microsoft/onnxruntime/blob/rel-1.16.1/cmake/deps.txt)?

**Comment by mc-nv** [comment]
@snnn , @pranavsharma for viz.

**Comment by snnn** [comment]
The fix will be in 1.16.2 release.

**Comment by snnn** [comment]
Related change: #15875

**Comment by syassami** [comment]
I made the following change locally (Apple Silicon) 

```
./build.sh --build_wheel --config Release --parallel --skip_submodule_sync --compile_no_warning_as_error --skip_submodule_sync --apple_deploy_target 12
```


```diff
diff --git a/cmake/deps.txt b/cmake/deps.txt
index 1e685511b..9bf0bb081 100644
--- a/cmake/deps.txt
+++ b/cmake/deps.txt
@@ -42,4 +42,4 @@ safeint;https://github.com/dcleblanc/SafeInt/archive/ff15c6ada150a5018c5ef217240
 tensorboard;https://github.com/tensorflow/tensorboard/archive/373eb09e4c5d2b3cc2493f0949dc4be6b6a45e81.zip;67b833913605a4f3f499894ab11528a702c2b381
 cutlass;https://github.com/NVIDIA/cutlass/archive/refs/tags/v3.0.0.zip;0f95b3c1fc1bd1175c4a90b2c9e39074d1bccefd
 extensions;https://github.com/microsoft/onnxruntime-extensions/archive/94142d8391c9791ec71c38336436319a2d4ac7a0.zip;4365ac5140338b4cb75a39944a4be276e3829b3c
-eigen;https://gitlab.com/libeigen/eigen/-/archive/3.4/eigen-3.4.zip;ee201b07085203ea7bd8eb97cbcb31b07cfa3efb
+eigen;https://gitlab.com/libeigen/eigen/-/archive/3.4/eigen-3.4.zip;3f6556b9666ced3c1f6fae389218c55c06f93096
```

Otherwise would run into 

```
onnxruntime/build/MacOS/Release/_deps/eigen-src/Eigen/src/Core/PartialReduxEvaluator.h:57:64: error: no matching function for call to 'pset1'
PacketType packetwise_redux_empty_value(const Func& ) { return pset1<PacketType>(0); }
                                                               ^~~~~~~~~~~~~~~~~
onnxruntime/build/MacOS/Release/_deps/eigen-src/Eigen/src/Core/PartialReduxEvaluator.h:112:14: note: in instantiation of function template specialization 'Eigen::internal::packetwise_redux_empty_value<__attribute__((neon_vector_type(8))) __fp16, Eigen::internal::scalar_sum_op<Eigen::half>>' requested here
      return packetwise_redux_empty_value<PacketType>(func);
```


Caused by 

```
onnxruntime/onnxruntime/contrib_ops/cpu/inverse.cc:62:19: note: in instantiation of function template specialization 'Eigen::Map<Eigen::Matrix<Eigen::half, -1, -1, 1>>::operator=<Eigen::Inverse<Eigen::Map<const Eigen::Matrix<Eigen::half, -1, -1, 1>>>>' requested here
    output_matrix = input_matrix.inverse();
```

**Comment by snnn** [comment]
@syassami  see PR #18308.

**Comment by devigned** [comment]
@snnn, I'm not clear how the #18308 would fix main. I've opened #18322 to track this build failure. I'll attempt to update the version of the eigen dep to see if that will fix the build failure on darwin/arm64.

**Comment by snnn** [comment]
The problem is fixed in onnxruntime 1.16.2 patch release.
